File: src/bot/commands/music/spotify/track.ts
/**
 * track.ts -- Retrieves information on the specified Spotify track.
 *
 * Copyright (c) 2019-present Kamran Mackey.
 *
 * Ellie is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ellie is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ellie. If not, see <https://www.gnu.org/licenses/>.
 */

import { Command } from 'discord-akairo';
import { Message, MessageEmbed } from 'discord.js';
import Constants from '../../../utils/Constants';

const moment = require('moment');
require('moment-duration-format');

export default class TrackCommand extends Command {
  public constructor() {
    super('spotify-track', {
      category: 'Music',
      clientPermissions: ['EMBED_LINKS'],
      description: {
        content: 'Displays information about the specified Spotify track.',
        usage: '<track>',
      },
      args: [
        {
          id: 'track',
          match: 'content',
          type: 'string',
        },
      ],
    });
  }

  public async exec(message: Message, { track }: { track: string }) {
    const trackEmbed = new MessageEmbed().setColor(0x1DB954);
    const errorEmbed = new MessageEmbed().setColor(0xB00020);

    if (!track) {
      errorEmbed.setTitle('Error: No track name provided.');
      errorEmbed.setDescription('You did not provide the name of the track you would '
        + 'like to get information on. Please provide one and then try again.');

      return message.channel.send(errorEmbed);
    }

    this.client.spotify.clientCredentialsGrant().then((data) => {
      this.client.spotify.setAccessToken(data.body.access_token);
      this.client.spotify.searchTracks(track, { limit: 1, offset: 0 }, (err, res) => {
        if (err) {
          console.log(err);
          return message.channel.send('Looks like an error occured while searching for a Spotify ' +
            'track. Please try again later.!');
        }

        try {
          const track = res.body.tracks.items[0];
          const parent = res.body.tracks.items[0].album; // The album belonging to the track.
          const title = track.name;
          const album = parent.name;
          const albumUrl = parent.external_urls.spotify;
          const popularity = track.popularity;
          const explicit = track.explicit ? 'Yes' : 'No';
          const cover = parent.images[1].url;
          const markets = track.available_markets.length;
          const url = track.external_urls.spotify;
          const date = moment(track.album.release_date).format('LL');
          const artist = track.artists.map(a => `[${a.name}](${a.external_urls.spotify})`).join(', ');
          const length = moment.duration(track.duration_ms, 'milliseconds').format('h[hr] mm[m] s[s]');

          // Set basic embed variables before setting description.
          trackEmbed.setTitle(title);
          trackEmbed.setThumbnail(cover);

          this.client.spotify.getAudioAnalysisForTrack(track.id).then((res) => {
            const analysis = res.body.track;
            const key = Constants.MUSIC_PITCH_CLASSES[analysis.key];
            const keyConfidence = analysis.key_confidence;
            const loudness = analysis.loudness;
            const mode = Constants.MUSIC_MODES[analysis.mode];
            const modeConfidence = analysis.mode_confidence;
            const tempo = analysis.tempo;
            const tempoConfidence = analysis.tempo_confidence;
            const timeSignature = analysis.time_signature;
            const timeSignatureConfidence = analysis.time_signature_confidence;

            trackEmbed.setDescription(
              `**Artist(s)**: ${artist}\n` +
              `**Album**: [${album}](${albumUrl})\n` +
              `**Popularity**: ${popularity}\n` +
              `**Explicit?** ${explicit}\n` +
              `**Release Date**: ${date}\n` +
              `**Spotify Markets**: ${markets}\n` +
              `**Duration**: ${length}\n\n` +
              '**__Technical Details__**:\n' +
              `**Average Loudness**: ${loudness} dB\n` +
              `**Key**: ${key} (Confidence: ${keyConfidence})\n` +
              `**Mode**: ${mode} (Confidence: ${modeConfidence})\n` +
              `**Tempo**: ${tempo} (Confidence: ${tempoConfidence})\n` +
              `**Time Signature**: ${timeSignature} (Confidence: ${timeSignatureConfidence})\n\n` +
              `[Play ${title} on Spotify →](${url})`);

            return message.channel.send(trackEmbed);
          });
        } catch {
          return message.channel.send('Sorry, I was unable to find that. Perhaps you made a typo? ' +
            'Or Spotify lacks the track you were searching for. Please try again later.');
        }
      });
    });
  }
}

File: src/commands/music/spotify/commands/track.rs
use chrono::prelude::*;

use humantime::format_duration;

use itertools::Itertools;

use std::time::Duration;

use serenity::client::Context;
use serenity::framework::standard::macros::command;
use serenity::framework::standard::Args;
use serenity::framework::standard::{CommandError, CommandResult};
use serenity::model::prelude::Message;

use crate::spotify;

#[command]
#[description("Displays information about a specified track on Spotify.")]
fn track(ctx: &mut Context, msg: &Message, args: Args) -> CommandResult {
    if args.rest().is_empty() {
        return msg
            .channel_id
            .send_message(&ctx, move |m| {
                m.embed(move |e| {
                    e.title("Error: No track name provided.");
                    e.description(
                        "You did not provide a track name. Please enter one and \
                        then try again.",
                    );
                    e
                })
            })
            .map_or_else(|e| Err(CommandError(e.to_string())), |_| Ok(()));
    }

    let track_name = args.rest();
    let track_search = spotify().search_track(&track_name, 1, 0, None);
    let track_result = &track_search.unwrap().tracks.items;
    
    let track = track_result.first().unwrap();
    let track_id = &track.id.clone().unwrap();
    let track_album = track.album.clone();
    let track_name = &track.name;
    let track_album_name = &track_album.name;
    let track_album_url = &track_album.external_urls["spotify"];
    let track_date = track_album.release_date.unwrap();
    let track_explicit = match track.explicit {
        true => "Yes".to_owned(),
        false => "No".to_owned()
    };
    let track_popularity = &track.popularity;
    let track_position = &track.track_number;
    let track_preview_url = track.preview_url.as_ref().unwrap();
    let track_markets = &track.available_markets.len();
    let track_length = format_duration(Duration::from_millis(track.duration_ms as u64 / 1000 * 1000));
    let track_url = &track.external_urls["spotify"];
    let track_image = &track_album.images.first().unwrap().url;
    let track_artists = &track.artists.iter().map(|a| format!("[{}]({})", &a.name, &a.external_urls["spotify"])).join(", ");
    let track_date = NaiveDate::parse_from_str(&track_date, "%Y-%m-%d").map_or(track_date, move |d| {
        let formatted_string = d.format("%B %-e, %Y").to_string();
        format!("{}", formatted_string.trim())
    });

    let track_analysis = spotify().audio_analysis(track_id).unwrap().track;

    let track_key = match track_analysis.key {
        0 => "C".to_owned(),
        1 => "C♯, D♭".to_owned(),
        2 => "D".to_owned(),
        3 => "D♯".to_owned(),
        4 => "E".to_owned(),
        5 => "F".to_owned(),
        6 => "F♯, G♭".to_owned(),
        7 => "G".to_owned(),
        8 => "G♯, A♭".to_owned(),
        9 => "A".to_owned(),
        10 => "A♯, B♭".to_owned(),
        11 => "B".to_owned(),
        _ => track_analysis.key.to_string(),
    };

    let track_key_confidence = track_analysis.key_confidence;
    let track_loudness = track_analysis.loudness;
    let track_mode = match track_analysis.mode as u32 {
        0 => "Minor".to_owned(),
        1 => "Major".to_owned(),
        _ => track_analysis.mode.to_string() 
    };
    let track_mode_confidence = track_analysis.mode_confidence;
    let track_tempo = track_analysis.tempo;
    let track_tempo_confidence = track_analysis.tempo_confidence;
    let track_time_signature = track_analysis.time_signature;
    let track_time_signature_confidence = track_analysis.time_signature_confidence;

    msg.channel_id.send_message(&ctx, move |m| {
        m.embed(move |e| {
            e.title(track_name);
            e.url(track_url);
            e.color(0x1DB954);
            e.thumbnail(track_image);
            e.description(format!(
                "\
                **Artist(s)**: {}\n\
                **Album**: [{}]({}) (track #{})\n\
                **Popularity**: {}\n\
                **Explicit?**: {}\n\
                **Release date**: {}\n\
                **Play preview**: [click here]({})\n\
                **Markets**: {}\n\
                **Length**: {}\n\
                **Loudness**: {} dB\n\
                **Key**: {} (confidence: {})\n\
                **Mode**: {} (confidence: {})\n\
                **Tempo**: {} (confidence: {})\n\
                **Time Signature**: {} (confidence: {})\n\n\
                [Play {} on Spotify →]({})
                ", 
                track_artists, track_album_name, track_album_url, track_position, 
                track_popularity, track_explicit, track_date, track_preview_url, track_markets, 
                track_length, track_loudness, track_key, track_key_confidence, track_mode, 
                track_mode_confidence,track_tempo, track_tempo_confidence, track_time_signature, 
                track_time_signature_confidence, track_name, track_url
            ));
            e.footer(|f| f.text("Powered by the Spotify Web API."))
        })
    }).map_or_else(|e| Err(CommandError(e.to_string())), |_| Ok(()))
}
